app/javascript/packs/document-capture.jsx

```diff
@@ -141,34 +141,34 @@ loadPolyfills(['fetch', 'crypto']).then(async () => {
 
   render(
     <DeviceContext.Provider value={device}>
-      <AcuantContextProvider
```

Comment:
These changes only move up the analytics provider. It's easier to see using "Hide whitespace changes" setting in GitHub:

![image](https://user-images.githubusercontent.com/1779930/112843806-fc62c380-9070-11eb-8a5c-e187de82578f.png)

Reply:
Related: Following a prior similar comment about moving to some flattened, "composed" provider implementation, I explored this a bit. I think it may help improve diffs in review, though I think the general readability does suffer a bit, at least in the explored implementation.

```jsx
function ComposeComponents({ components, children }) {
  return components.reduceRight(
    (result, [Provider, props]) => (
      // eslint-disable-next-line react/jsx-props-no-spreading
      <Provider {...props}>{result}</Provider>
    ),
    children
  );
}
```

```jsx
render(
  <ComposeComponents
    components={[
      [DeviceContext.Provider, { value: device }],
      [AnalyticsContext.Provider, { value: { addPageAction, noticeError } }],
      [
        AcuantContextProvider,
        {
          credentials: getMetaContent("acuant-sdk-initialization-creds"),
          endpoint: getMetaContent("acuant-sdk-initialization-endpoint"),
        },
      ],
      [
        UploadContextProvider,
        {
          endpoint: /** @type {string} */ (appRoot.getAttribute(
            "data-endpoint"
          )),
          statusEndpoint: /** @type {string} */ (appRoot.getAttribute(
            "data-status-endpoint"
          )),
          statusPollInterval:
            Number(appRoot.getAttribute("data-status-poll-interval-ms")) ||
            undefined,
          method: isAsyncForm ? "PUT" : "POST",
          csrf,
          isMockClient,
          backgroundUploadURLs,
          backgroundUploadEncryptKey,
          formData,
        },
      ],
      [I18nContext.Provider, { value: i18n.strings }],
      [ServiceProviderContext.Provider, { value: getServiceProvider() }],
      [AssetContext.Provider, { value: assets }],
    ]}
  >
    <DocumentCapture isAsyncForm={isAsyncForm} onStepChange={keepAlive} />
  </ComposeComponents>,
  appRoot
);
```

Reply:
I agree -- a little diff noise here and there seems easier than trying to mentally unpack the loop